torch/csrc/distributed/rpc/python_functions.cpp

```diff
@@ -32,17 +32,25 @@ py::object toPyObj(const Message& message) {
       Stack stack;
       stack.push_back(ret.value());
       {
-        pybind11::gil_scoped_acquire ag;
-        // The createPyObjectForStack does not acquire GIL, but creating a new
-        // py::object requires GIL.
-        return torch::jit::createPyObjectForStack(std::move(stack));
+        // Need GIL to guard createPyObjectForStack() and its returned
+        // py::object
+        py::gil_scoped_acquire acquire;
+        return jit::toIValue(
+            torch::jit::createPyObjectForStack(std::move(stack)),
+            PyObjectType::get());
       }
     }
     case MessageType::PYTHON_RET: {
       // TODO: Try to avoid a copy here.
       auto& resp = static_cast<PythonResp&>(*response);
       auto& pythonRpcHandler = PythonRpcHandler::getInstance();
-      return pythonRpcHandler.deserialize(resp.serializedPyObj());
+      {
+        // Need GIL to destruct the py::object returned by deserialize()
+        py::gil_scoped_acquire acquire;
+        return jit::toIValue(
+            pythonRpcHandler.deserialize(resp.serializedPyObj()),
```

Comment:
It works because `py::gil_scoped_acquire` is constructed before the temp `py::object` is returned by `pythonRpcHandler.deserialize(..)`, they are destructed in reverse order.